### PR TITLE
Move clusterIP and clusterDomain spec fields to status

### PR DESCRIPTION
**Pull request by pravisankar**
https://jira.coreos.com/browse/NE-160

### CODE REVIEW

**Comment by pravisankar** [comment]
@openshift/sig-network-edge PTAL

**Comment by Miciah** [comment]
Initializing status in the factory method seems wrong: the responsibility of the factory should be to provide a resource with metadata and spec, and it should be up to the reconciliation logic (perhaps `ensureCoreDNSForClusterDNS`, or some new method that the `reconcile` loop would invoke) to update the status; does that seem reasonable?

In addition, what do you think about enabling the `/status` subresource on the CRD in order to get correct semantics for updates to the resource metadata and spec versus updates to its its status?

File: pkg/apis/dns/v1alpha1/types.go [hunk 1]
```diff
@@ -22,10 +25,13 @@ type ClusterDNS struct {
 }
 
 type ClusterDNSSpec struct {
+}
+
+// ClusterDNSStatus defines the observed state of ClusterDNS
+type ClusterDNSStatus struct {
+	// ClusterIP is the service IP reserved for cluster DNS service
 	ClusterIP *string `json:"clusterIP"`
```
Comment (ironcladlou) [inline pkg/apis/dns/v1alpha1/types.go]: Should these still be pointers?

File: pkg/manifests/manifests.go [hunk 2]
```diff
@@ -196,8 +174,8 @@ func (f *Factory) DNSService(dns *dnsv1alpha1.ClusterDNS) (*corev1.Service, erro
 	}
 	s.Spec.Selector["dns"] = s.Name
 
-	if dns.Spec.ClusterIP != nil {
-		s.Spec.ClusterIP = *dns.Spec.ClusterIP
+	if len(clusterIP) > 0 {
+		s.Spec.ClusterIP = clusterIP
```
Comment (ironcladlou) [inline pkg/manifests/manifests.go]: I must be missing something, didn't `Spec` get deleted? surprised this compiles

File: pkg/stub/handler.go [hunk 3]
```diff
@@ -175,7 +166,19 @@ func (h *Handler) ensureDNSNamespace() error {
...
-	ds, err := h.ManifestFactory.DNSDaemonSet(dns)
+	clusterIP := ""
+	// TODO: fetch this from higher level openshift resource when it is exposed
+	clusterDomain := "cluster.local"
+
+	if dns.Name == "default" {
+		var err error
+		clusterIP, err = getClusterIPFromNetworkConfig()
+		if err != nil {
+			return fmt.Errorf("failed to getch cluster IP from network config for clusterdns %s, %v", dns.Name, err)
```
Comment (Miciah) [inline pkg/stub/handler.go]: Typo: "getch".

File: pkg/stub/handler.go [hunk 4]
```diff
@@ -175,7 +166,19 @@ func (h *Handler) ensureDNSNamespace() error {
 // ensureCoreDNSForClusterDNS ensures all necessary CoreDNS resources exist for
 // a given clusterdns.
 func (h *Handler) ensureCoreDNSForClusterDNS(dns *dnsv1alpha1.ClusterDNS) error {
-	ds, err := h.ManifestFactory.DNSDaemonSet(dns)
+	clusterIP := ""
+	// TODO: fetch this from higher level openshift resource when it is exposed
+	clusterDomain := "cluster.local"
+
+	if dns.Name == "default" {
```
Comment (Miciah) [inline pkg/stub/handler.go]: I'm ambivalent about special-casing on the clusterdns's name.  For now, clusterdns is a singleton, right? Does it make sense to special-case the name now? Will it make sense to special-case the name in the future, if we expand the API to support multiple clusterdnses?

File: pkg/stub/handler.go [hunk 5]
```diff
@@ -240,3 +244,58 @@ func (h *Handler) ensureDNSDeleted(dns *dnsv1alpha1.ClusterDNS) error {
...
+// syncClusterDNSStatus updates the status for a given clusterdns.
+func syncClusterDNSStatus(dns *dnsv1alpha1.ClusterDNS, clusterIP, clusterDomain string) error {
+	dns.Status = dnsv1alpha1.ClusterDNSStatus{
+		ClusterIP:     clusterIP,
+		ClusterDomain: clusterDomain,
+	}
+
+	unstructObj, err := k8sutil.UnstructuredFromRuntimeObject(dns)
+	if err != nil {
+		return fmt.Errorf("failed to get unstructed obj from runtime: %v", err)
```
Comment (Miciah) [inline pkg/stub/handler.go]: Typo: "unstructed" should be "unstructured".  (In https://github.com/openshift/cluster-dns-operator/pull/76/commits/2f85300f92b76f3723b5fb7b01d3ba136aff0e34#diff-f33ebabd902db5bc0032360e4c24c3c8R97, I used the phrasing "failed to convert"—not sure what phrasing most useful in the logs.)

File: pkg/stub/handler.go [hunk 6]
```diff
@@ -240,3 +244,58 @@ func (h *Handler) ensureDNSDeleted(dns *dnsv1alpha1.ClusterDNS) error {
...
+			Name: "cluster",
+		},
+	}
+	err := sdk.Get(networkConfig)
+	if err != nil {
+		return "", fmt.Errorf("failed to get network 'cluster': %v", err)
+	}
+
+	if len(networkConfig.Status.ServiceNetwork) == 0 {
+		return "", fmt.Errorf("no service networks found in cluster network config")
```
Comment (Miciah) [inline pkg/stub/handler.go]: :+1:, adding this check is good.

File: pkg/manifests/manifests.go (hunk 2 above)
Comment (Miciah) [reply pkg/manifests/manifests.go]: That's setting `.spec.clusterIP` on the service, not the clusterdns.

**Comment by pravisankar** [comment]
@openshift/sig-network-edge ready for review/merge, PTAL

**Pull request merged**